app/controllers/api/base_controller/parser.rb

```diff
@@ -331,6 +323,35 @@ def assert_all_required_fields_exists(data, type, required_fields)
           raise BadRequestError, "Resource #{missing_fields.join(", ")} needs be specified for creating a new #{type}"
         end
       end
+
+      def validate_provider_class
+        param = params['provider_class']
+        return unless param.present?
+
+        raise BadRequestError, "Unsupported provider_class #{param} specified" if param != "provider"
+        %w(tags policies policy_profiles).each do |cname|
+          if @req.subcollection == cname || @req.expand?(cname)
+            raise BadRequestError, "Management of #{cname} is unsupported for the Provider class"
+          end
+        end
+        @collection_klasses[:providers] = Provider
+      end
+
+      def validate_collection_class
+        param = params['collection_class']
+        return unless param.present?
+
+        klass = collection_class(@req.collection)
+        return if param == klass.name
+
+        param_klass = klass.descendants.detect { |sub_klass| param == sub_klass.name }
```

Comment:
Yeah, that's usually a big no-no, which was why I was commenting.  Glad you found a much nicer alternative!